test/controllers/products_controller_test.rb

```diff
@@ -26,13 +26,66 @@
     end
   end
 
-  it "must get the show view" do
-    get product_path(one.id)
-    must_respond_with :success
+  describe "show" do
+    it "must get the show view" do
+      get product_path(one.id)
+      must_respond_with :success
+
+      get product_path(-1)
+      must_respond_with :redirect
+      must_redirect_to root_path
+    end
+  end
+
+  describe "new product" do
+    it "should get new if a mercant is signed in" do
+      @merchant = merchants(:sappy1)
+      login(@merchant, :github)
+      get new_product_path
+      must_respond_with :success
+    end
+
+    it "must redirect to root path if a merchant is not signed in" do
+      # With no mercant signed in
+      get edit_product_path(one.id)
+      must_respond_with :redirect
+      flash[:status].must_equal :error
```

Comment:
it might be a good idea to test the flash[:result_text] as well to ensure we get the correct error.